### DISCUSSION TITLE
Doubts in policy.xml

### DISCUSSION THREAD

**Discussion by programmerarunkumar** [ImageMagick/ImageMagick#4212] [Help] [answered]
policy domain="resource" name="temporary-path" value="/data/magick"

what is the use of temporary-path property ?
Do we need to create folder for imagemagick data ?

When this folder will be used ?
It is the place where imagemagick will store data for processing ?

How long the data will be persisted in that ImageMagickData Folder
If it is persisted for long time  can we delete that data after certain time ?

**Comment by snibgo** [comment]
When IM does its work, it stores pixel values in structures called "pixel caches". If there is enough free memory, that is where they are stored. Otherwise, they are written to disk files in the temporary-path, with filenames that start with "magick". IM should delete these files when they are no longer needed, but if IM is killed by an external process, it doesn't get the chance to delete them.

IM doesn't use those files between commands. You can delete them whenever you want. If a "magick*" file is currently being used by IM, ordinary users won't be able to delete them.

It is good housekeeping practise to delete temporary files, every day or every week or whatever.

**Reply by programmerarunkumar** [reply]
> If there is enough free memory, that is where they are stored. Otherwise, they are written to disk files in the temporary-path, with filenames that start with "magick"

**that is where they are stored**
can you please elaborte this I don't understand?

**Comment by snibgo** [comment, marked as answer]
When IM reads an image file, it decodes the file, which often involves decompressing the pixels. IM stores the pixel values in a structure that is something like a 2-dimensional array, so there is fast access to any of the pixels.

But the image may have 100 billion pixels, and your computer may have only 10 GB of memory. When this happens, IM can't store the pixels in memory. So it stores them in files on a disk instead.

**Comment by programmerarunkumar** [comment]
If Memory(Heap) limit is exceeded , we will store image pixel data in Memory Map
If Memory Map limit is exceeded , we will store image pixel data  in disk 
What will happen when disk limit is exceeded ?
image process will be slow ?

**Comment by snibgo** [comment]
> What will happen when disk limit is exceeded ?

The IM program cannot continue so it will fail, sending an error message to stderr.

**Comment by programmerarunkumar** [comment]
thread -> Number of threads that are permitted to run in parallel for one image or overall ?

**Comment by snibgo** [comment]
That is the limit on the number of concurrent threads for that process. IM doesn't process images in parallel. The multi-threading is for processing of each image.

**Comment by programmerarunkumar** [comment]
What is better value for Thread count 
When below code is called Maximum of 5times at a time 
NewMagickWand();
MagickReadImage();
MagickResizeImage();
MagickWriteImages();
DestroyMagickWand();

**Comment by snibgo** [comment]
In that list, only MagickResizeImage() may use multi-threading. However, it may use OpenCL. You should use your O/S tools to determine if multi-threading occurs. On Windows, taskmgr.exe show CPU utilisation. 100% means that all available thread are being used.

For example, if your CPU has 8 cores so can run 8 threads at once, and MagickResizeImage() uses 8 threads, then doing 5 resizes simultaneously won't work well, so limiting each IM multithreading to 1 or 2 may be more efficient (lower overall time to complete the tasks).

**Reply by programmerarunkumar** [reply]
> For example, if your CPU has 8 cores so can run 8 threads at once, and MagickResizeImage() uses 8 threads, then doing 5 resizes simultaneously won't work well, so limiting each IM multithreading to 1 or 2 may be more efficient (lower overall time to complete the tasks). 

 Why doing 5 resizes simultaneously won't work well if we use the thread count same as cpu cores count ?

**Comment by snibgo** [comment]
When you start your code "5 times at a time", each will take one thread. Then each resize will try to use the maximum thread count, 8 threads in my example. So you will have 5 * 8 = 40 threads running at the same time, on a CPU that has only 8 cores. Most threads will spend most of their time waiting for a time slot. The housekeeping of starting, pausing, re-starting and ending threads may consume more time than the actual resizing.

This also depends on the amount of memory you have available, and how much each of your "5 times at a time" needs. Trying to use more memory than you have available will certainly decrease performance (because disk will be used instead) and may cause program failure (if IM needs memory for something that can't be cached to disk).

For best overall performance, you need to do your own testing. My processor can have 8 threads running at the same time, but best performance comes from using fewer threads, such as 4.

**Comment by programmerarunkumar** [comment]
Approximately what percentage of cpu cores can be used as thread count ?​

**Comment by snibgo** [comment]
You need to do your own testing.

**Comment by urban-warrior** [comment]
See https://imagemagick.org/script/openmp.php.

**Comment by programmerarunkumar** [comment]
Thanks for the clarification
